frame/support/src/storage/mod.rs

```diff
@@ -563,6 +572,103 @@ impl<T> Iterator for PrefixIterator<T> {
 	}
 }
 
+/// Iterate over a prefix of a child trie and decode raw_key and raw_value into `T`.
+///
+/// If any decoding fails it skips the key and continues to the next one.
+pub struct ChildTriePrefixIterator<T> {
+	prefix: Vec<u8>,
+	/// Storage key of the child trie
+	storage_key: Vec<u8>,
+	previous_key: Vec<u8>,
+	/// If true then values are removed while iterating
+	drain: bool,
+	/// Function that takes `(raw_key_without_prefix, raw_value)` and decode `T`.
+	/// `raw_key_without_prefix` is the raw storage key without the prefix iterated on.
+	closure: fn(&[u8], &[u8]) -> Result<T, codec::Error>,
+}
+
+impl<T> ChildTriePrefixIterator<T> {
+	/// Mutate this iterator into a draining iterator; items iterated are removed from storage.
+	pub fn drain(mut self) -> Self {
+		self.drain = true;
+		self
+	}
+}
+
+impl<T: Decode + Sized> ChildTriePrefixIterator<(Vec<u8>, T)> {
+	/// Construct iterator to iterate over child trie items in `child_info` with the prefix `prefix`.
```

Comment:
some interesting not is that the prefix iterator doesn't fetch the value on the prefix itself.

Like if there is a value at `prefix` this value is not the first value returned by the prefix iterator, it is never returned by the prefix iterator.

Maybe we can fix it by adding a field: `prefix_fetched: bool` which would be a false when iterator is created.
We can do that for both main trie and child trie iterator.

Reply:
I noticed from the iterator algorithm -- it always skips the first value under `prefix`, and I thought this is intended behavior. Is there a chance that there might be a value stored under `prefix`?

Reply:
the iterator is usually used to fetch value in a storage map. In this context there is no key under the prefix.
But I think it is a bit unexpected behavior if ppl start using in other context.

At some point we can add `prefix_fetched: bool` and add a method to fetch or not fetch the prefix itself similarly to the drain method.

Reply:
sorry my suggestion is mixing multiple stuff together, what I am thinking it that:

* `with_storage` function name is not very clear because usual pallet storages are stored in the main trie. So maybe better to just rename it `iter_prefix` which just iterate on prefix.
* if we rename `iter_prefix` then either we say in the doc that the value on prefix is not fetched or (and I think it is better) we also fetch the value on the prefix. To do so we modify slightly ChildTriePrefixIterator with a new field `prefix_fetched: bool` and use it to fetch the prefix or not when iterating.